File: apps/engine-heritage/src/engine_heritage/digest.py
"""sha256 of the canonical outcome (PRD §6.2) — mirrors
Domain/Games/Engine/BlackRed/Digest.php's pipe-joined format, not JSON canonicalisation."""

from __future__ import annotations

import hashlib


def of(seed_hex: str, selected_positions: list[int], board: list[int], winning_positions: list[int], outcome_tier: str) -> str:
    parts = [
        seed_hex,
        ",".join(str(p) for p in selected_positions),
        ",".join(str(n) for n in board),
        ",".join(str(p) for p in winning_positions),
        outcome_tier,
    ]
    return hashlib.sha256("|".join(parts).encode("ascii")).hexdigest()

File: apps/platform/app/Domain/Games/Engine/BlackRed/Digest.php
<?php

declare(strict_types=1);

namespace App\Domain\Games\Engine\BlackRed;

/** §6.2 — sha256 of the canonical outcome, returned alongside every resolve/replay. */
final class Digest
{
    /**
     * @param list<'B'|'R'> $prediction
     * @param list<'B'|'R'> $result
     */
    public static function of(string $seedHex, array $prediction, array $result): string
    {
        return hash('sha256', $seedHex . '|' . implode('', $prediction) . '|' . implode('', $result));
    }
}
